glean-core/rlb/src/dispatcher/mod.rs

```diff
@@ -247,3 +273,288 @@ impl Dispatcher {
+        enable_test_logging();
+
+        let thread_canary = Arc::new(AtomicU8::new(0));
+
+        let mut dispatcher = Dispatcher::new(100);
+
+        // Add 3 tasks to queue each one increasing thread_canary by 1 to
+        // signal that the tasks ran.
+        for _ in 0..3 {
+            let canary_clone = thread_canary.clone();
+            dispatcher
+                .guard()
+                .launch(move || {
+                    canary_clone.fetch_add(1, Ordering::SeqCst);
+                })
+                .expect("Failed to dispatch the test task");
+        }
+
+        // Ensure that no task ran.
+        assert_eq!(0, thread_canary.load(Ordering::SeqCst));
+
+        // Flush the queue and wait for the tasks to complete.
+        dispatcher
+            .flush_init()
+            .expect("Failed to get out of preinit queue mode");
+        dispatcher.block_on_queue();
+        // Validate that we have the expected canary value.
+        assert_eq!(3, thread_canary.load(Ordering::SeqCst));
+    }
+
+    #[test]
+    fn queued_tasks_are_flushed_off_the_main_thread() {
+        enable_test_logging();
+
+        let main_thread_id = thread::current().id();
+        let thread_canary = Arc::new(AtomicU8::new(0));
+
+        let mut dispatcher = Dispatcher::new(100);
+
+        // Add 3 tasks to queue each one increasing thread_canary by 1 to
+        // signal that the tasks ran.
+        for _ in 0..3 {
+            let canary_clone = thread_canary.clone();
+            dispatcher
+                .guard()
+                .launch(move || {
+                    assert!(thread::current().id() != main_thread_id);
+
+                    canary_clone.fetch_add(1, Ordering::SeqCst);
+                })
+                .expect("Failed to dispatch the test task");
+        }
+
+        // Ensure that no task ran.
+        assert_eq!(0, thread_canary.load(Ordering::SeqCst));
+
+        // Flush the queue and wait for the tasks to complete.
+        dispatcher
+            .flush_init()
+            .expect("Failed to get out of preinit queue mode");
+        dispatcher.block_on_queue();
+        // Validate that we have the expected canary value.
+        assert_eq!(3, thread_canary.load(Ordering::SeqCst));
+    }
+
+    #[test]
+    fn tasks_are_processed_in_order() {
+        enable_test_logging();
+
+        let mut dispatcher = Dispatcher::new(10);
+        dispatcher.flush_init().unwrap();
+
+        let result = Arc::new(Mutex::new(vec![]));
+        for i in 1..=5 {
+            let result = Arc::clone(&result);
+            dispatcher
+                .guard()
+                .launch(move || {
+                    result.lock().unwrap().push(i);
+                })
+                .unwrap();
+        }
+
+        dispatcher.block_on_queue();
+
+        assert_eq!(&*result.lock().unwrap(), &[1, 2, 3, 4, 5]);
+    }
+
+    #[test]
+    fn preinit_tasks_are_processed_after_flush() {
```

Comment:
This test is a superset of the above test, is it not? We can probably do without the smaller test.